02.ComputeShader/main.cpp

```diff
@@ -43,11 +50,40 @@ class ComputeShaderSampleApp : public ApplicationBase
 
 	core::smart_refctd_ptr<video::IGPUComputePipeline> m_pipeline = nullptr;
 	core::vector<core::smart_refctd_ptr<video::IGPUDescriptorSet>> m_descriptorSets;
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> m_descriptorSetLayout;
 
 	// These can be removed after descriptor lifetime tracking
 	core::vector<core::smart_refctd_ptr<video::IGPUImageView>> m_swapchainImageViews;
 	core::smart_refctd_ptr<video::IGPUImageView> m_inImageView;
 
+	core::deque<CommonAPI::IRetiredSwapchainResources*> m_qRetiredSwapchainResources;
+	std::mutex m_resizeLock;
+
+	uint32_t m_swapchainIteration = 0;
+	std::array<uint32_t, CommonAPI::InitOutput::MaxSwapChainImageCount> m_imageSwapchainIterations;
+	nbl::video::ISwapchain::SCreationParams m_swapchainCreationParams;
+
+	struct CSwapchainResources : public CommonAPI::IRetiredSwapchainResources
+	{
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> oldImageView = nullptr;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImage> oldImage = nullptr;
+
+		~CSwapchainResources() override {}
+	};
+
+	struct ResizeEventHandler : public CommonAPI::CommonAPIEventCallback
```

Comment:
0. Let's get rid of `ICommonAPIEventCallback` because it's redundant and used in only 1 place to introduce `setLogger`
1. let's rename this class to `EventHandler` because we might decide to handle other events in the example with it.

With our current event handling design we need an instance of the event handler in order to get the events, the default one worked well until we needed to handle some events such as resize on our own.

Now the problem statement is "How can the event handler access the members of our example class" i.e the relation between the two.

## Current Solution
the current ugly solution right now is the `OnResize` class and the example inheriting from it to implement the virtual function.
In this case, What would we do if we wanted to specialize some other events like `onWindowRotated` or `onWindowMaximized`? Do we create new classes and inherit from them? Or do we have a single class `OnEvents` and have all the functions there and example class inherits from `OnEvents`? (may be actually a solution but not a good one)
We are in need of a solution that connects event handling to examples.

## Solution 1 - examples inherit `CommonAPIEventHandler`
Let the examples that use window inherit from `CommonAPIEventHandler` and then user passes the event handler itself in `InitParams` instead of letting CommonAPI construct it for them.
this way user can override the callbacks it needs.

I'm not so sure about this one because it kinda has dependancy on logger and input system which common apu provides for us, but we can simply have `setter` functins for them.

IF we make it work This solution would be convenient and better than the currentnt one and we wouldn't even need to have ~~Resize~~EventHandler 

## Solution 2 - register std::function
Our examples don't inherit from anything but ~~Resize~~EventHandler contains a bunch of `std::function`s with correct signatures like `std::function<void(uint32_t,uint32_t)>`

unless there is some weird reason not to use a std::function this is a pretty good choice.

## Solution 3 - the event functions in ApplicationFramework

This solution has the GraphicalApplication (or maybe IGraphicalApplicationFramework) define virtual functions like `OnResize`, and the `CommonAPI::CommonAPIEventCallback` stores a `IGraphicalApplicationFramework*` and has a function like `setIGraphicalApplication` and we can so something like:

```c++
bool onWindowResized_impl(uint32_t w, uint32_t h) override
{
	m_logger.log("Window resized to { %u, %u }", nbl::system::ILogger::ELL_DEBUG, w, h);
        if(graphicalApp)
           graphicalApp->onResize(w,h);
	return true;
}
```

I like this solution, but we're re-defining some function like `OnResize` similar to `onWindowResized_impl` again as opposed to Solution#1 where we straight up inherit from `CommonAPI::CommonAPIEventCallback` and the example will be it's own event handler.

the problem with Solution#1 is having to decide at compile time whether our example will need window and events but It's not a big deal.

We can combine this with Solution#1 and Have `GraphicalApplication inherit ICommonAPIEventCallback` or even one step further and `nbl::ui::IGraphicalApplicationFramework inherit nbl::ui::IWindow::IEventCallback`
It has the same downside of Solution#1 where we have to move the callback creation out of `CommonAPI::Init `and probably put it in `InitParams`

Let me know if you have other solutions in mind. we need to come to a conclusion about this.